lib/eibflatpak.py

```diff
@@ -344,17 +351,36 @@ def reset_metadata(self):
             remote.set_title(self.title)
         if self.default_branch:
             remote.set_default_branch(self.default_branch)
+        if self.prio is not None:
+            remote.set_prio(self.prio)
+        if self.noenumerate is not None:
+            remote.set_noenumerate(self.noenumerate)
         self.installation.modify_remote(remote)
 
+    @contextmanager
+    def _allow_enumerate(self):
+        if not self.noenumerate:
+            yield
+        else:
+            remote = self.installation.get_remote_by_name(self.name)
+            try:
+                remote.set_noenumerate(False)
+                self.installation.modify_remote(remote)
+                yield
+            finally:
+                remote.set_noenumerate(self.noenumerate)
+                self.installation.modify_remote(remote)
```

Comment:
Is the problem here that calling `list_remote_refs_sync` on a remote with `noenumerate = true` fails? Another option is to treat `noenumerate` specially and handle it in the `deploy` method where it would only get set at the end.

Reply:
Oh, I like that idea! I ended up with this diff, although I need to test it on Monday before I push a commit somewhere. I don't think I properly understand what `reset_metadata` is for, so I might have broken something here :)

```diff
diff --git a/lib/eibflatpak.py b/lib/eibflatpak.py
index 7aed1ad..4870be4 100644
--- a/lib/eibflatpak.py
+++ b/lib/eibflatpak.py
@@ -238,8 +238,6 @@ class FlatpakRemote(object):
             remote.set_default_branch(self.default_branch)
         if self.prio is not None:
             remote.set_prio(self.prio)
-        if self.noenumerate is not None:
-            remote.set_noenumerate(self.noenumerate)
 
         # Import the GPG key if specified
         if self.gpg_key:
@@ -303,12 +301,7 @@ class FlatpakRemote(object):
         Sync the appstream and metadata from the remote. If the deploy
         URL differs from the pull URL, it's adjusted here, too.
         """
-        if self.deploy_url != self.url:
-            logger.info('Setting %s URL to %s for deployment', self.name,
-                        self.deploy_url)
-            remote = self.installation.get_remote_by_name(self.name)
-            remote.set_url(self.deploy_url)
-            self.installation.modify_remote(remote)
+        self._modify_remote_for_deployment()
 
         # Set the flatpak remote collection ID if it's enabled and the
         # remote has a collection ID
@@ -335,15 +328,37 @@ class FlatpakRemote(object):
         # Reset any configuration defined metadata
         self.reset_metadata()
 
+    def _modify_remote_for_deployment(self):
+        if self.deploy_url == self.url and self.noenumerate is None:
+            return
+
+        remote = self.installation.get_remote_by_name(self.name)
+
+        if self.deploy_url != self.url:
+            logger.info('Setting %s URL to %s for deployment', self.name,
+                        self.deploy_url)
+            remote.set_url(self.deploy_url)
+        
+        if self.noenumerate is not None:
+            logger.info('Setting %s noenumerate to %s for deployment',
+                        self.name, self.noenumerate)
+            remote.set_noenumerate(self.noenumerate)
+
+        self.installation.modify_remote(remote)
+
     def reset_metadata(self):
         """Reset remote's metadata per configuration
 
-        Set the title and default branch remote options back to the
+        Set, title, default branch, and prio remote options back to the
         value used in the builder configuration. This is needed if the
-        setting was changed by a call to update the remote's metadata
-        from the server.
+        setting was changed by a call to update the remote's metadata from
+        the server.
         """
-        if not self.title and not self.default_branch:
+        if (
+            not self.title
+            and not self.default_branch
+            and self.prio is None
+        ):
             return
 
         remote = self.installation.get_remote_by_name(self.name)
@@ -353,24 +368,8 @@ class FlatpakRemote(object):
             remote.set_default_branch(self.default_branch)
         if self.prio is not None:
             remote.set_prio(self.prio)
-        if self.noenumerate is not None:
-            remote.set_noenumerate(self.noenumerate)
         self.installation.modify_remote(remote)
 
-    @contextmanager
-    def _allow_enumerate(self):
-        if not self.noenumerate:
-            yield
-        else:
-            remote = self.installation.get_remote_by_name(self.name)
-            try:
-                remote.set_noenumerate(False)
-                self.installation.modify_remote(remote)
-                yield
-            finally:
-                remote.set_noenumerate(self.noenumerate)
-                self.installation.modify_remote(remote)
-
     def enumerate(self):
         """Populate refs from remote data
 
@@ -378,9 +377,8 @@ class FlatpakRemote(object):
         app or runtime found.
         """
         logger.info('Fetching refs for %s', self.name)
-        with self._allow_enumerate():
-            all_remote_refs = eib.retry(
-                self.installation.list_remote_refs_sync, self.name)
+        all_remote_refs = eib.retry(
+            self.installation.list_remote_refs_sync, self.name)
         for remote_ref in all_remote_refs:
             # Get the full ostree ref
             ref = remote_ref.format_ref()
```